### ISSUE TITLE
Socket close on Timeout

### ISSUE THREAD

**Issue by tmds** [aspnet/KestrelHttpServer#1771]
I am looking in the code to see how the Timeout is implemented.
I think something is missing that unblocks the socket read or I missed it.

Pseudo code of socket reading:
```
do {
buffer = socket.read();                   // (1)
completed = pipewriter.flush(buffer);
} while (!completed);
pipewriter.complete();
socket.shutdown(readend);                 // (2)
```

On timeout we want to reach (2) this requires getting unblocked at (1).

**Comment by davidfowl** [comment]
When a timeout is triggered, the request processing loop completes, which completes the output reader. When that happens the socket transports then shuts down both sides of the connection https://github.com/aspnet/KestrelHttpServer/blob/e76737b8a63d0230feae1a493d81be8b676387bf/src/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets/SocketConnection.cs#L58. Then read yields with a 0 byte read and ends the connection.

**Comment by tmds** [comment]
Ah, when the application completes the output reader, the Transport is responsible for unblocking the socket reading. Got it, Thanks.

**Comment by tmds** [comment]
I wonder, would it be possible for a client to block the timeout logic by not reading?
Pseudo code of socket writing:

```
do {
buffer = pipereader.read();
if (buffer.iscomplete) break;
socket.send(buffer);           // (1)
} while(true);
pipereader.complete();
socket.dispose();              // (2)
```

That is: (1) stops you from getting to (2)?

**Comment by davidfowl** [comment]
Yes, it's possible the request processing ends but socket itself is stuck on a slow client call. You'll be able to get a notification out of band with this change https://github.com/dotnet/corefxlab/pull/1491.

**Comment by muratg** [comment]
@blowdart Thoughts on this one

**Comment by blowdart** [comment]
The ability for a client to block by not reading is kind of bad to me. It needs to happen regardless.

**Comment by cesarbs** [comment]
This is fixed on the libuv transport now: #1939

Does "socket" in this issue's title refer to sockets in general or to the socket transport specifically? There's a new issue filed to get the write timeout working on the Sockets transport (#1943), so I think we could close this one.

**Comment by tmds** [comment]
@cesarbs indeed, this can be closed then.
